services/application.go

```diff
@@ -92,3 +95,23 @@ func (app *ChainlinkApplication) AddJob(job models.JobSpec) error {
 	app.Scheduler.AddJob(job)
 	return app.EthereumListener.AddJob(job, app.HeadTracker.LastRecord())
 }
+
+// AddAdapter adds an adapter to the store. If another
+// adapter with the same name already exists the adapter
+// will not be added.
+func (app *ChainlinkApplication) AddAdapter(bt *models.BridgeType) error {
+	store := app.GetStore()
+
+	app.bridgeTypeMutex.Lock()
+	defer app.bridgeTypeMutex.Unlock()
+
+	if err := ValidateAdapter(bt, store); err != nil {
+		return utils.NewValidationError(err.Error())
```

Comment:
One idea is to place the `ValidationError` inside `services/validators.go`. I actually believe we need to differentiate between errors in the method above, `AddJob` as well. But we'll save that for another PR.

Reply:
When I thought of putting errors in the services package it didn't feel right to me since an error is not a a service.

Reply:
The thinking is that errors are themselves not necessarily a service, but they are part of the related logic so we put them in the same package. This makes it easier to encapsulate all the required parts in one package. We already have somewhat of a [precedent for this](https://github.com/smartcontractkit/chainlink/blob/master/services/job_runner.go#L157).

Reply:
Have you guys thought about differentiating between interface, model, and controller logic? Throughout the code I'm seeing files that have interfaces, funcs, structs ,etc  all in the same file if they are somewhat related - this causes the files to be fairly long and to break the single responsibility principle. Placing the error in a folder structure that contains only model logic would make more sense to me.

Reply:
> The single responsibility principle is a computer programming principle that states that every module or class should have responsibility over a single part of the functionality provided by the software, and that responsibility should be entirely encapsulated by the class.

[Wikipedia](https://en.wikipedia.org/wiki/Single_responsibility_principle)

The single responsibility principle (SRP) is usually applied to types or classes, not files. Just because there are many interfaces and types in a file does not mean they are breaking SRP. You could place an MVC web service in one file, and still adhere to SRP. This wouldn't be the case if you had a single coupled `main` method.

A trend in frameworks such as React is to place all relevant parts of a component in a single file so there's one place to look. This is different to what you propose, which is prevalent in ruby and rails where there's usually [one file per type](https://github.com/rails/rails/blob/master/activejob/lib/active_job/serializers.rb).

Idiomatic `go` has many types in the same file, similar to python, partially because of the package abstraction. You can see [go/scan.go right here](https://github.com/golang/go/blob/master/src/fmt/scan.go#L148-L152) has a lot going on, including an error type, but the types inside still adhere to SRP. 

In our code base, `ValidationError` has one responsibility, but that doesn't break just because it's in a file with other methods. We actually started off only having one type per file, especially in the [`adapters`](https://github.com/smartcontractkit/chainlink/tree/master/adapters) folder, but ending up following what we saw in idiomatic go, which resulted in the [`http.go`](https://github.com/smartcontractkit/chainlink/blob/master/adapters/http.go) file holding multiple adapters.

As for this coupling between interface, model, and controller logic, our models are explicitly split from our controllers, hence the package `models` and the controllers in `web`.  When you say uncoupling are you speaking just about file and package layout and was that covered?

Let me know what you think, I'm all for design discussions. Thanks for staying diligent @aklintsevich.

Reply:
I'm not familiar with Go design patterns. If this is the accepted and recommended way to go for Go programming then it makes more sense.

Reply:
Yeah, it takes some getting used to, but that seems to be [the accepted way to write idiomatic go](https://stackoverflow.com/questions/26228839/golang-code-organisation-with-structs-variables-and-interfaces). Having said that, if you think there are any types that are too tightly coupled in logic, let us know.